packages/common/src/failure.ts

```diff
@@ -128,11 +145,44 @@ export class ApplicationFailure extends TemporalFailure {
    * @param type Optional error type
    * @param details Optional details about the failure. Serialized by the Worker's {@link PayloadConverter}.
    */
-  public static nonRetryable(message: string | undefined, type?: string, ...details: unknown[]): ApplicationFailure {
+  public static nonRetryable(
+    message?: string | undefined | null,
+    type?: string | undefined | null,
+    ...details: unknown[]
+  ): ApplicationFailure {
     return new this(message, type ?? 'Error', true, details);
   }
 }
 
+interface ApplicationFailureOptions {
+  /**
+   * Error message
+   */
+  message?: string;
+
+  /**
+   * Error type (used by {@link RetryPolicy.nonRetryableErrorTypes})
+   */
+  type?: string;
+
+  /**
+   * Whether the current Activity or Workflow can be retried
+   *
+   * @default false
+   */
+  nonRetryable?: boolean;
+
+  /**
+   * Details about the failure. Serialized by the Worker's {@link PayloadConverter}.
+   */
+  details?: unknown[];
+
+  /**
+   * Cause of the failure
+   */
+  cause?: Error;
```

Comment:
This can only be a TemporalFailure for it to be properly serialized.

Reply:
I think a common use case is putting a 3rd party library's `Error` here. The message and stackTrace should get serialized fine: https://github.com/temporalio/sdk-typescript/blob/32c12fd2aaec1acf6866cfe0b47c3893c428cc15/packages/common/src/failure.ts#L337-L344

Reply:
:+1: And `TemporalFailure` accepts `cause` as non-Temporal error. Python `from error` which we support, Go https://pkg.go.dev/go.temporal.io/sdk/temporal#NewApplicationErrorWithCause, and Java https://www.javadoc.io/static/io.temporal/temporal-sdk/1.14.0/io/temporal/failure/ApplicationFailure.html#newFailureWithCause(java.lang.String,java.lang.String,java.lang.Throwable,java.lang.Object...) all accept generic causes

Reply:
True, thanks